### PR TITLE
Refactor backend for incremental rebuilds

### LINKED ISSUE #212: Implement incremental rebuild
(Rebuild only when source/dependencies modified.)
I've made a start on this, so creating an issue to assign.

### PR DESCRIPTION
**Pull request by LKedward**
This PR contains the required changes to implement incremental rebuilds (Resolves #212):
- Moves enumeration of link objects to occur before the backend is called
- Adds source file hashing to determine whether targets need rebuilding
- Separates target scheduling from target compilation which:
  - allows unit-testing of backend sorting and scheduling
  - optionally allows parallel builds when compiled with `openmp` (bootstrap: `fpm build --flag -fopenmp --flag -O3`)

### CODE REVIEW

File: fpm/src/fpm_backend.f90 [hunk 1]
```diff
@@ -1,113 +1,203 @@
...
+    do i=1,size(schedule_ptr)-1
 
+        ! Build targets in schedule region i
+        !$OMP PARALLEL DO DEFAULT(SHARED)
+        do j=schedule_ptr(i),(schedule_ptr(i+1)-1)
+
+            call build_target(model,queue(j)%ptr)
+
+        end do
+        !$OMP END PARALLEL DO
```
Comment (awvwgk) [inline fpm/src/fpm_backend.f90:51]: No `end parallel do` needed.

File: fpm/src/fpm_backend.f90 [hunk 2]
```diff
@@ -1,113 +1,203 @@
...
 
-end subroutine build_package
+    ! Construct build schedule queue
+    call schedule_targets(queue, schedule_ptr, model%targets)
 
+    ! Loop over parallel schedule regions
+    do i=1,size(schedule_ptr)-1
 
+        ! Build targets in schedule region i
+        !$OMP PARALLEL DO DEFAULT(SHARED)
```
Comment (awvwgk) [inline fpm/src/fpm_backend.f90:45]: Can we use lowercase OpenMP pragmas?

File: fpm/src/fpm_backend.f90 [hunk 3]
```diff
@@ -1,113 +1,203 @@
...
-recursive subroutine build_target(model,target,linking)
-    ! Compile Fortran source, called recursively on it dependents
-    !  
-    type(fpm_model_t), intent(in) :: model
-    type(build_target_t), intent(inout) :: target
-    character(:), allocatable, intent(in) :: linking
 
-    integer :: i, j, ilib
+!> Topologically sort a target for scheduling by 
+!>  recursing over it's dependencies.
```
Comment (awvwgk) [inline fpm/src/fpm_backend.f90:59]: Grammar:
```suggestion
!>  recursing over its dependencies.
```

**Review by awvwgk — APPROVED** [review]
Works really well so far. I had a few issues when switching branches with git and rebuilding, but those might also result from incompatible module files left in the build directory (I would clean the build directory anyway for those cases). So no real failures so far, which is great.

Just some minor comments below.

**Comment by LKedward** [comment]
Thanks for reviewing; suggestions applied.

> I had a few issues when switching branches with git and rebuilding, but those might also result from incompatible module files left in the build directory

I've been having this issue with the bootstrap version but can't find the source of the problem. I haven't (yet) encountered this problem with this Fortran implementation — I will look into it for a future PR :+1:

**Comment by awvwgk** [comment]
> I've been having this issue with the bootstrap version but can't find the source of the problem. I haven't (yet) encountered this problem with this Fortran implementation

Maybe I encountered this while running `fpm run -- build`, hard to tell exactly which fpm failed, since it was an issue from `gfortran` complaining about incompatible module files. I cleaned the build directory afterwards (`rm -r build/gfortran_*`), but we will certainly get the chance to mess the build process up often enough in the future.

**Review by milancurcic — APPROVED** [review]

**Comment by LKedward** [comment]
Many thanks for reviewing; I'll merge later today if there are no more comments.

**Review by urbanjost — APPROVED** [review]

**Comment by urbanjost** [comment]
Apparently there is an issue with an old version of Firefox working with the reviews. I thought I had approved this days ago but I noticed it is not checked. Others comments covered my issues so I have no problem with this going forward. Sorry if that caused a delay. Hopefully this works from this browser.

**Pull request merged**